src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -1362,6 +1332,103 @@ TEST(Aktualizr, DownloadListRemove) {
+      aktualizr.SetSignalHandler(std::bind(targets_autoclean_cb, std::ref(aktualizr), std::placeholders::_1));
+  aktualizr.Initialize();
+
+  {
+    result::UpdateCheck update_result = aktualizr.CheckUpdates().get();
+    aktualizr.Download(update_result.updates).get();
+
+    EXPECT_EQ(aktualizr.GetStoredTargets().size(), 1);
+
+    result::Install install_result = aktualizr.Install(update_result.updates).get();
+    EXPECT_TRUE(install_result.dev_report.success);
+
+    std::vector<Uptane::Target> targets = aktualizr.GetStoredTargets();
+    ASSERT_EQ(targets.size(), 1);
+    EXPECT_EQ(targets[0].filename(), "primary_firmware.txt");
+  }
+
+  // second install
+  repo.emptyTargets();
+  repo.addImage(fake_meta_dir / "fake_meta/dummy_firmware.txt", "dummy_firmware.txt", hwid, "", {});
+  repo.addTarget("dummy_firmware.txt", hwid, "CA:FE:A6:D2:84:9D", "");
+  repo.signTargets();
+
+  {
+    result::UpdateCheck update_result = aktualizr.CheckUpdates().get();
+    aktualizr.Download(update_result.updates).get();
+
+    EXPECT_EQ(aktualizr.GetStoredTargets().size(), 2);
+
+    result::Install install_result = aktualizr.Install(update_result.updates).get();
+    EXPECT_TRUE(install_result.dev_report.success);
+
+    // all targets are kept (current and previous)
+    std::vector<Uptane::Target> targets = aktualizr.GetStoredTargets();
+    ASSERT_EQ(targets.size(), 2);
+  }
+
+  // third install (first firmware again)
+  repo.emptyTargets();
+  repo.addImage(fake_meta_dir / "fake_meta/primary_firmware.txt", "primary_firmware.txt", hwid, "", {});
+  repo.addTarget("primary_firmware.txt", hwid, "CA:FE:A6:D2:84:9D", "");
+  repo.signTargets();
+
+  {
+    result::UpdateCheck update_result = aktualizr.CheckUpdates().get();
+    aktualizr.Download(update_result.updates).get();
+
+    EXPECT_EQ(aktualizr.GetStoredTargets().size(), 2);
+
+    result::Install install_result = aktualizr.Install(update_result.updates).get();
+    EXPECT_TRUE(install_result.dev_report.success);
+
+    // all targets are kept again (current and previous)
+    std::vector<Uptane::Target> targets = aktualizr.GetStoredTargets();
+    ASSERT_EQ(targets.size(), 2);
+  }
+
+  // fourth install (some new third firmware)
+  repo.emptyTargets();
+  repo.addImage(fake_meta_dir / "fake_meta/secondary_firmware.txt", "secondary_firmware.txt", hwid, "", {});
+  repo.addTarget("secondary_firmware.txt", hwid, "CA:FE:A6:D2:84:9D", "");
+  repo.signTargets();
+
+  {
+    result::UpdateCheck update_result = aktualizr.CheckUpdates().get();
+    aktualizr.Download(update_result.updates).get();
+
+    EXPECT_EQ(aktualizr.GetStoredTargets().size(), 3);
+
+    result::Install install_result = aktualizr.Install(update_result.updates).get();
+    EXPECT_TRUE(install_result.dev_report.success);
+
+    // only two targets are left: dummy_firmware has been cleaned up
+    std::vector<Uptane::Target> targets = aktualizr.GetStoredTargets();
+    ASSERT_EQ(targets.size(), 2);
```

Comment:
This is a pretty awesome test, but do you think it's worth checking that the two remaining are the two we expect at the very end?

Reply:
Yes, I'll add that.